src/source/Crypto/Dtls_openssl.c

```diff
@@ -104,26 +104,48 @@ STATUS createCertificateAndKey(INT32 certificateBits, BOOL generateRSACertificat
 {
     ENTERS();
     STATUS retStatus = STATUS_SUCCESS;
-    BIGNUM* pBne = NULL;
-    RSA* pRsa = NULL;
     X509_NAME* pX509Name = NULL;
-    UINT32 eccGroup = 0;
-    EC_KEY* eccKey = NULL;
     UINT64 certSn;
 
     CHK(ppCert != NULL && ppPkey != NULL, STATUS_NULL_ARG);
     CHK((*ppPkey = EVP_PKEY_new()) != NULL, STATUS_CERTIFICATE_GENERATION_FAILED);
     CHK_STATUS(dtlsFillPseudoRandomBits((PBYTE) &certSn, SIZEOF(UINT64)));
 
+#if (OPENSSL_VERSION_NUMBER >= 0x30000000L)
+    DLOGI("Detected openssl version greater than 3.0.0");
+    EVP_PKEY_CTX* pctx = NULL;
+    if (generateRSACertificate) {
+        DLOGI("Using RSA");
+        CHK_ERR(pctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL), STATUS_CERTIFICATE_GENERATION_FAILED, "Error creating EVP_PKEY_CTX for RSA");
+        CHK_ERR(EVP_PKEY_keygen_init(pctx) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "Error initializing RSA keygen");
+        CHK_ERR(EVP_PKEY_CTX_set_rsa_keygen_bits(pctx, certificateBits) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "Error setting RSA key size");
+        CHK_ERR(EVP_PKEY_keygen(pctx, ppPkey) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "RSA Key generation errored");
+    } else {
+        DLOGI("Using EC");
+        CHK_ERR(pctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL), STATUS_CERTIFICATE_GENERATION_FAILED, "Error creating EVP_PKEY_CTX for EC");
+        CHK_ERR(EVP_PKEY_keygen_init(pctx) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "Error initializing EC keygen");
+        CHK_ERR(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, OBJ_sn2nid("prime256v1")) > 0, STATUS_CERTIFICATE_GENERATION_FAILED,
+                "Error setting curve name");
+        CHK_ERR(EVP_PKEY_keygen(pctx, ppPkey) > 0, STATUS_CERTIFICATE_GENERATION_FAILED, "EC Key generation errored");
+    }
+    EVP_PKEY_CTX_free(pctx);
+#else
+    RSA* pRsa = NULL;
     if (generateRSACertificate) {
+        BIGNUM* pBne = NULL;
```

Comment:
Defining this here instead of at the start breaks the C89 / old compiler compliance we have enforced in our SDK.  These need to be moved back up to the top of the function.

Reply:
Hmm. The reason I added it here was because Mac was complaining about unused variables. I can move it on the top and #if #else there too then.